subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java

```diff
@@ -121,6 +125,7 @@ protected void compile(IncrementalTaskInputs inputs) {
             getPath(),
             (IncrementalTaskInputsInternal) inputs,
             source,
+            getSource(),
```

Comment:
Can we make this work without passing both `getSource()` and the raw `source`? It seems like we can just use `getSource()` and detect root directories in the (composite) tree, no? You can call `FileCollectionInternal.visitrRootElements()` perhaps?

Reply:
Switched to using `visitRootElements` instead.